File: src/main/java/gildedrose/GildedRose.java
package gildedrose;

import java.util.Arrays;

public class GildedRose {
    private final Item[] items;

    public GildedRose(Item[] items) {
        this.items = Arrays.stream(items)
                .map(GildedRose::createSpecializedItem)
                .toArray(Item[]::new);
    }

    public Item[] getItems() {
        return items;
    }

    private static Item createSpecializedItem(Item item) {
        if (item.getClass() != Item.class) {
            return item;
        }
        return Item.create(item.getName(), item.getSellIn(), item.getQuality());
    }

    public void updateQuality() {
        for (Item item : items) {
            item.updateQuality();
        }
    }
}

File: src/gildedrose/gilded_rose.py
from gildedrose.item import Item


class GildedRose:
    """Runs the nightly quality/sell_in update across a shop's inventory."""

    def __init__(self, items: list[Item]) -> None:
        self.items = items

    def update_quality(self) -> None:
        for item in self.items:
            item.update()
